rust/sbp/src/link.rs

```diff
@@ -0,0 +1,369 @@
+        msg.try_into().unwrap()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{cell::RefCell, sync::mpsc::channel, thread, time::Duration};
+
+    use crate::messages::{
+        self,
+        observation::{MsgObs, MsgObsDepA, ObservationHeader, ObservationHeaderDep},
+    };
+
+    use super::*;
+
+    #[test]
+    fn test_link_triggered() {
+        let source = LinkSource::new();
+        let link = source.link();
+        let triggered = RefCell::new(false);
+        link.register(|triggered: &RefCell<bool>, _: SBP| {
+            *triggered.borrow_mut() = true;
+        });
+        source.send_state(&triggered, SBP::from(make_msg_obs()));
+        assert!(*triggered.borrow());
+    }
+
+    #[test]
+    fn test_link_multi_thd() {
+        let (s, r) = channel();
+
+        let source = LinkSource::new();
+        let link = source.link();
+        let triggered = RefCell::new(false);
+
+        let handle = thread::spawn(move || {
+            link.register(|triggered: &RefCell<bool>, _: SBP| {
+                *triggered.borrow_mut() = true;
+            });
+            s.send(()).unwrap();
+        });
+        r.recv_timeout(Duration::from_secs(1)).unwrap();
+
+        source.send_state(&triggered, SBP::from(make_msg_obs()));
+
+        handle.join().unwrap();
+
+        assert!(*triggered.borrow());
+    }
+
+    #[test]
+    fn test_link_no_state() {
+        let mut triggered = false;
+        {
+            let source = LinkSource::new();
+            let link = source.link();
+            link.register(|_: SBP| {
+                triggered = true;
+            });
+            source.send(SBP::from(make_msg_obs()));
+        }
+        assert!(triggered);
+    }
+
+    #[test]
+    fn test_link_remove() {
+        let source = LinkSource::new();
+        let link = source.link();
+        let count = RefCell::new(0);
+
+        let key = link.register(|count: &RefCell<usize>, _: SBP| {
+            *count.borrow_mut() += 1;
+        });
+
+        let msg = SBP::from(make_msg_obs());
+        source.send_state(&count, &msg);
+        assert_eq!(*count.borrow(), 1);
+
+        link.unregister(key);
+        source.send_state(&count, &msg);
+        assert_eq!(*count.borrow(), 1);
+    }
+
+    #[test]
+    fn test_custom_event() {
+        enum ObsMsg {
+            Obs(MsgObs),
+            DepA(MsgObsDepA),
+        }
+
+        impl Event for ObsMsg {
+            const MESSAGE_TYPES: &'static [u16] = &[MsgObs::MESSAGE_TYPE, MsgObsDepA::MESSAGE_TYPE];
+            fn from_sbp(msg: SBP) -> Self {
+                match msg {
+                    SBP::MsgObs(m) => ObsMsg::Obs(m),
+                    SBP::MsgObsDepA(m) => ObsMsg::DepA(m),
+                    _ => unreachable!("wrong event keys"),
+                }
+            }
+        }
+
+        let source = LinkSource::new();
+        let link = source.link();
+        let count = RefCell::new(0);
+
+        link.register(|count: &RefCell<usize>, _: SBP| {
```

Comment:
Should this test receiving an ObsMsg struct?

Reply:
Yup good catch